dspace-oai/src/main/java/org/dspace/xoai/app/XOAI.java

```diff
@@ -162,52 +162,132 @@ public int index() throws DSpaceSolrIndexerException {
+            // Index both in_archive items AND withdrawn items. Withdrawn items
+            // will be flagged withdrawn
             // (in order to notify external OAI harvesters of their new status)
-            Iterator<Item> iterator = itemService.findInArchiveOrWithdrawnDiscoverableModifiedSince(
-                    context, null);
-            return this.index(iterator);
+            Iterator<Item> discoverableItems = itemService.findInArchiveOrWithdrawnDiscoverableModifiedSince(context,
+                    null);
+            Iterator<Item> nonDiscoverableItems = itemService
+                    .findInArchiveOrWithdrawnNonDiscoverableModifiedSince(context, null);
+            return this.index(discoverableItems) + this.index(nonDiscoverableItems);
         } catch (SQLException ex) {
             throw new DSpaceSolrIndexerException(ex.getMessage(), ex);
         }
     }
 
-    private int index(Iterator<Item> iterator)
-            throws DSpaceSolrIndexerException {
+    /**
+     * Check if an item is already indexed. Using this, it is possible to check
+     * if withdrawn or nondiscoverable items have to be indexed at all.
+     * 
+     * @param item
+     *            Item that should be checked for its presence in the index.
+     * @return has it been indexed?
+     */
+    private boolean checkIfIndexed(Item item) {
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.id");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            return documents.getNumFound() == 1;
+        } catch (DSpaceSolrException | SolrServerException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Check if an item is flagged visible in the index.
+     * 
+     * @param item
+     *            Item that should be checked for its presence in the index.
+     * @return has it been indexed?
+     */
+    private boolean checkIfVisibleInOAI(Item item) {
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.public");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            if (documents.getNumFound() == 1) {
+                return (boolean) documents.get(0).getFieldValue("item.public");
+            } else {
+                return false;
+            }
+        } catch (DSpaceSolrException | SolrServerException e) {
+            return false;
+        }
+    }
+
+    private int index(Iterator<Item> iterator) throws DSpaceSolrIndexerException {
         try {
             int i = 0;
             SolrServer server = solrServerResolver.getServer();
             while (iterator.hasNext()) {
                 try {
                     Item item = iterator.next();
-                    server.add(this.index(item));
+                    System.out.println("indexing item " + item.getID());
```

Comment:
I'm worried this will get noisy in production, as if you have 100K items to index, you are going to see this message 100K times.